clang/test/SemaSYCL/union-kernel-param-neg.cpp

```diff
@@ -0,0 +1,53 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -fcxx-exceptions -verify -fsyntax-only %s
+
+// This test checks if compiler reports compilation error on an attempt to pass
+// accessor/sampler/pointer as SYCL kernel parameter inside union.
+
+#include "sycl.hpp"
+using namespace cl::sycl;
+
+union union_with_sampler {
+  cl::sycl::sampler smpl;
+  // expected-error@-1 {{'cl::sycl::sampler' cannot be used as the type of a kernel parameter}}
```

Comment:
Is this error message sufficient?  Should we point out that it is 'in a union'?  Perhaps add a select to the end of the error messsage?

Reply:
@erichkeane . The diagnostic message needs to be improved. As we talked about it, this will be done in a separate PR.

Reply:
I am seeing strange buildbot error on windows. Build is fine on my linux workspace.

I am not sure why it is failing. 

FAILED: tools/clang/lib/Sema/CMakeFiles/obj.clangSema.dir/SemaSYCL.cpp.obj 
C:\PROGRA~2\MIB055~1\2017\PROFES~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x64\cl.exe  /nologo /TP -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GNU_SOURCE -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\clang\lib\Sema -ID:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\clang\lib\Sema -ID:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\llvm\..\llvm-spirv\include -ID:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\clang\include -Itools\clang\include -Iinclude -ID:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\llvm\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:__cplusplus /Zc:strictStrings /Oi /Zc:rvalueCast /bigobj /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd4324 -w14062 -we4238 /Gw /MD /O2 /Ob2    /EHs-c- /GR- -UNDEBUG -std:c++14 /bigobj /showIncludes /Fotools\clang\lib\Sema\CMakeFiles\obj.clangSema.dir\SemaSYCL.cpp.obj /Fdtools\clang\lib\Sema\CMakeFiles\obj.clangSema.dir\ /FS -c D:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\clang\lib\Sema\SemaSYCL.cpp
D:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\clang\lib\Sema\SemaSYCL.cpp(985): error C2660: '`anonymous-namespace'::KernelObjVisitor::VisitUnion': function does not take 5 arguments
D:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\clang\lib\Sema\SemaSYCL.cpp(889): note: see declaration of '`anonymous-namespace'::KernelObjVisitor::VisitUnion'
D:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\clang\lib\Sema\SemaSYCL.cpp(2176): note: see reference to function template instantiation 'void `anonymous-namespace'::KernelObjVisitor::VisitRecordFields<`anonymous-namespace'::SyclKernelFieldChecker,`anonymous-namespace'::SyclKernelUnionChecker>(const clang::CXXRecordDecl *,`anonymous-namespace'::SyclKernelFieldChecker &,`anonymous-namespace'::SyclKernelUnionChecker &)' being compiled
D:\buildbot\Product_worker_intel\sycl-win-x64-pr\llvm.src\llvm\include\llvm/Support/TrailingObjects.h(135): note: see reference to class template instantiation 'llvm::trailing_objects_internal::TrailingObjectsImpl<8,BaseTy,TopTrailingObj,NextTy>' being compiled
        with
        [

Reply:
FYI, the 'error message' question is more the wording of this individual message.  What we talked about doing in a different patch was printing the error 'backtrace'.

I think we want to change the error message itself for unions to be more explicit.

Reply:
Oh No!  That seems that MSVC has a bug that isn't working with my visit union :/  It doesn't have a great error message either.

Reply:
The problems seems here: Each of the VisitUnion calls now has the FilteredHandlers type FIRST instead of right at the end in their param lists. It was added to catch missing flags. MSVC seems complaining about this during instantiation since VisitRecord adds this at the end.

Reply:
> FYI, the 'error message' question is more the wording of this individual message. What we talked about doing in a different patch was printing the error 'backtrace'.
> 
> I think we want to change the error message itself for unions to be more explicit.

I will change the error message.

Reply:
> The problems seems here: Each of the VisitUnion calls now has the FilteredHandlers type FIRST instead of right at the end in their param lists. It was added to catch missing flags. MSVC seems complaining about this during instantiation since VisitRecord adds this at the end.

That part should still work... its an empty pack.  I tried moving it to the back, to the front and I'm not sure the filtering works, have you confirmed that it still works moving it to the front?

Reply:
MSVC is a little stupid with empty packs: https://godbolt.org/z/od3r4q

See the added 2 definitions of VisitUnion

Reply:
Fixed this. Thanks Erich.